test-adapter/src/main/scala/org/scalajs/testing/adapter/HTMLRunnerBuilder.scala

```diff
@@ -26,85 +26,70 @@ import org.scalajs.testing.common._
 
 /** Template for the HTML runner. */
 object HTMLRunnerBuilder {
-
-  private val tmpSuffixRE = """[a-zA-Z0-9-_.]*$""".r
-
-  private def tmpFile(path: String, in: InputStream): URI = {
-    try {
-      /* - createTempFile requires a prefix of at least 3 chars
-       * - we use a safe part of the path as suffix so the extension stays (some
-       *   browsers need that) and there is a clue which file it came from.
-       */
-      val suffix = tmpSuffixRE.findFirstIn(path).orNull
-
-      val f = File.createTempFile("tmp-", suffix)
-      f.deleteOnExit()
-      Files.copy(in, f.toPath(), StandardCopyOption.REPLACE_EXISTING)
-      f.toURI()
-    } finally {
-      in.close()
-    }
-  }
-
   def writeToFile(output: File, title: String, input: Seq[Input],
       frameworkImplClassNames: List[List[String]],
       taskDefs: List[TaskDef]): Unit = {
+    writeToFile(output.toPath, title, input, frameworkImplClassNames, taskDefs)
+  }
 
-    val jsFiles = input.map {
-      case Input.Script(script) => script
-
-      case _ =>
-        throw new UnsupportedInputException(
-            s"Unsupported input for the generation of an HTML runner: $input")
-    }
+  def writeToFile(output: Path, title: String, input: Seq[Input],
+      frameworkImplClassNames: List[List[String]],
+      taskDefs: List[TaskDef]): Unit = {
 
-    val jsFileURIs = jsFiles.map { f =>
+    def tryRelativize(content: Path): Option[Path] = {
       try {
-        f.toFile.toURI
+        Some(output.relativize(content))
       } catch {
-        case _: UnsupportedOperationException =>
-          tmpFile(f.toString, Files.newInputStream(f))
+        case _: IllegalArgumentException => None
       }
     }
 
-    val cssURI = {
-      val name = "test-runner.css"
-      tmpFile(name, getClass.getResourceAsStream(name))
+    def scriptTag(tpe: String, content: Path) = {
+      tryRelativize(content).fold {
+        s"""
+        <script type="$tpe">
+          ${new String(Files.readAllBytes(content), UTF_8)}
+        </script>
+        """
+      } { relPath =>
+        s"""<script type="$tpe" src="${htmlEscaped(relPath.toString)}"></script>"""
```

Comment:
Unfortunately, this simple strategy will not honor the guarantee of executing all the `Input`s in the specified order. The problem is that ES modules are executed in `defer` mode by default (and this cannot be changed, other than to `async`) whereas inline scripts and non-module src scripts are executed in synchronous mode by default.

Reference: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script

So if the input is like
```
List(ESModule(file1), Script(file2))
```
with the above strategy, `file2` will be executed first, and `file1` afterwards.

We can force non-module src scripts to execute in `defer` mode with the `defer` attribute of the `<script>` tag. So that would easily solve the problem as long as there are no inline script.

However, inline scripts cannot be executed in `defer` mode. AFAICT, if the input is like my example and `file2` is emitted as an inline script, there does not exist any combination of option that will force execution of `file1` before `file2`.

So, unfortunately, I think that prevents us from using inline scripts in the general case. We have to fall back to emitting temp files.

Reply:
> We have to fall back to emitting temp files.

Hmm... But then we have the problem of where to put them :-/ If modules are used, everything needs to be served via HTTP, otherwise the browser simply disallows it.

So either we move everything to tmp, maybe even serving it ourselves from sbt.

Maybe the best option is to require a target directory for the HTML runner. That would then be the output with all the files in it (probably not even tmp at this point).

Reply:
> Maybe the best option is to require a target directory for the HTML runner. That would then be the output with all the files in it (probably not even tmp at this point).

Yes, that's probably the best option, indeed.